packages/tools/client-debugger/client-debugger-view/src/components/ContainerHistoryView.tsx

```diff
@@ -8,79 +8,68 @@ import React from "react";
+	HasContainerId,
+	ISourcedDebuggerMessage,
+	InboundHandlers,
 } from "@fluid-tools/client-debugger";
-
-import { HasClientDebugger } from "../CommonProps";
+import { useMessageRelay } from "../MessageRelayContext";
+import { Waiting } from "./Waiting";
 
 /**
- * {@link ContainerHistory} input props.
+ * {@link ContainerHistoryView} input props.
  */
-export type ContainerHistoryProps = HasClientDebugger;
+export type ContainerHistoryProps = HasContainerId;
 
 /**
- * Displays information about the provided {@link IFluidClientDebugger.getContainerConnectionLog}.
+ * Displays information about the container state history.
  *
  * @param props - See {@link ContainerHistoryViewProps}.
  */
 export function ContainerHistoryView(props: ContainerHistoryProps): React.ReactElement {
-	const { clientDebugger } = props;
-	const { container } = clientDebugger;
+	const { containerId } = props;
+	const messageRelay = useMessageRelay();
 
 	const [containerHistory, setContainerHistory] = React.useState<
 		readonly ConnectionStateChangeLogEntry[]
-	>(clientDebugger.getContainerConnectionLog());
+	>([]);
 
 	React.useEffect(() => {
-		function onContainerHistoryChanged(): void {
-			setContainerHistory(clientDebugger.getContainerConnectionLog());
+		/**
+		 * Handlers for inbound messages related to the registry.
+		 */
+		const inboundMessageHandlers: InboundHandlers = {
+			["CONTAINER_STATE_HISTORY"]: (untypedMessage) => {
+				const message = untypedMessage as ContainerStateHistoryMessage;
+				if (message.data.containerId === containerId) {
+					setContainerHistory(message.data.history);
+					return true;
+				}
+				return false;
+			},
+		};
+
+		/**
+		 * Event handler for messages coming from the webpage.
+		 */
+		function messageHandler(message: Partial<ISourcedDebuggerMessage>): void {
+			handleIncomingMessage(message, inboundMessageHandlers, {
+				context: "ContainerHistoryView", // TODO: Fix
+			});
 		}
 
-		container.on("attached", onContainerHistoryChanged);
-		container.on("connected", onContainerHistoryChanged);
-		container.on("disconnected", onContainerHistoryChanged);
-		container.on("disposed", onContainerHistoryChanged);
-		container.on("closed", onContainerHistoryChanged);
+		messageRelay.on("message", messageHandler);
 
 		return (): void => {
-			container.off("attached", onContainerHistoryChanged);
-			container.off("connected", onContainerHistoryChanged);
-			container.off("disconnected", onContainerHistoryChanged);
-			container.off("disposed", onContainerHistoryChanged);
-			container.off("closed", onContainerHistoryChanged);
+			messageRelay.off("message", messageHandler);
 		};
-	}, [clientDebugger, container, setContainerHistory]);
-
-	return (
-		<Stack
-			styles={{
-				root: {
-					height: "100%",
-				},
-			}}
-		>
-			<StackItem>
-				<h3>Container State History</h3>
-				<_ContainerHistoryView containerHistory={containerHistory} />
-			</StackItem>
-		</Stack>
-	);
-}
+	}, [containerId, messageRelay]);
```

Comment:
We still need `setContainerHistory` here, don't we?

Reply:
Yep! Good catch, thanks. While reading React docs I learned that the `set<>()` functions passed in to the component technically never change (they're the exact same function object) and thus never trigger effects even if they're listed as dependencies, so they're not _technically_ necessary, but we've been using them so I think we should keep them. If we make a conscious decision to stop putting them as deps everywhere, we can review that.